### DISCUSSION TITLE
Creating/Refreshing tokens as a recurring task

### DISCUSSION THREAD

**Discussion by ghost** [relayrides/pushy#960] [Q&A] [answered]
Hi there!

I've a couple of questions. APNs requires to refresh the auth token regularly and their documentation says it reports an error if the tokens are recreated more than once every 20 minutes.

1. How does Pushy address this scenario or refresh tokens?
2. If pushy is running in multiple container instances, it is possible each instance create tokens frequently. If I want to setup a recurring task to generate and store the token in database so that all instances share same token, how would I generate the token using this library? Is there an example?

Thanks for creating this!

Best,
Manideep

**Comment by jchambers** [comment, marked as answer]
> How does Pushy address this scenario or refresh tokens?

Each connection to the APNs server maintains internal state that tracks when its token was generated. It schedules a job to expire the token every 50 minutes. Please see [`TokenAuthenticationApnsClientHandler`](https://github.com/jchambers/pushy/blob/master/pushy/src/main/java/com/eatthepath/pushy/apns/TokenAuthenticationApnsClientHandler.java) for implementation details.

> If I want to setup a recurring task to generate and store the token in database so that all instances share same token, how would I generate the token using this library? Is there an example?

This is not necessary; token age restrictions are enforced at the level of individual connections rather than applications/sets of credentials. Because there's no need to synchronize tokens across connections, Pushy provides no affordance for doing so.

**Reply by ghost** [reply]
Thanks for your quick response!!
